File: code/ex922.java
class Solution {
    public int[] sortArrayByParityII(int[] A) {
        int n = A.length;
        int j = 1;
        for (int i = 0; i < n; i += 2) {
            if (A[i] % 2 == i % 2)    continue;
            
            while (A[j] % 2 != i % 2)    j += 2;
            // Swap A[i] with A[j]
            int temp = A[i];
            A[i] = A[j];
            A[j] = temp;
        }
        return A;
    }
}

File: code/ex922.cpp
class Solution {
public:
    vector<int> sortArrayByParityII(vector<int>& A) {
        int n = A.size();
        int odd = 0, even = 0, p = 0;
        while (p < n) {
            if (p % 2 == 0) {
                while (even < n && A[even] % 2 == 1)    even++;
                swap(A[p], A[even++]);
            } else {
                while (odd < n && A[odd] % 2 == 0)    odd++;
                swap(A[p], A[odd++]);
            }
            p++;
        }
        return A;
    }
};